### ISSUE TITLE
Nested comments led to syntax color not displayed

### ISSUE THREAD

**Issue by wangzhaohe** [leo-editor/leo-editor#3006]
**EKR**

This is not a bug. The `@doc` parts and Leo's syntax colorer are acting as expected.

I have hidden most of the comments because they concern other matters.

**Original Report**

The first three lines are right because html should not contain nested comments. But the syntax color is disappeared in leo-editor.

<img width="683" alt="image" src="https://user-images.githubusercontent.com/2628013/208216237-db7e999b-e568-4a9e-b814-7fd709be267f.png">

The last three lines has syntax color in leo-editor, but `-->` is shown in web browser.

<img width="573" alt="image" src="https://user-images.githubusercontent.com/2628013/208216564-c568a671-51d0-441e-ac4f-76e130424ba2.png">

**Comment by tbpassin** [comment]
I am not sure exactly what you are trying to write here, but your XML is not well-formed, because XML comments cannot be nested.  So far as the XML is concerned,  if you want to have what look like comments inside comments, you have to use _&gt;_ and _&lt;_ entities (or at least _&lt;_).  Also, inside a comment you cannot have the string "--", according the the RxP XML parser.

You could put unescaped special characters in a CDATA section, but I'm pretty sure that's not what you want.

Would you please attach a plain text file that contains the XML content that you want?  And then explain exactly what you are trying to do here.

**Comment by wangzhaohe** [comment]
It is a html file. I write @language xml because @language html will not show with VR3( will show the html render result).

**Comment by tbpassin** [comment]
Is this supposed to be an asciidoc file or an XML file?  And what are those _@doc_ directives intended to do?  VR3 does not do anything with them.

VR3's purpose is to show rendered versions of subtrees, including @file trees.  If you don't need to see rendered html, you don't need VR3, it seems to me.  With _@language xml_  VR3 does show unrendered but colorized text as you say.  But so does the body itself.

If this is an asciidoc file, then the Asciidoctor processor will render the markup the way it is programmed to do, and that is what you will see in VR3.

To sum up, I don't understand what you are trying to do, what you want to happen, and what is actually happening that is different from what you want.

Once I know those things, we can think about whether something can be done.

**Comment by tbpassin** [comment]
Oh, maybe I see what you are thinking of.  Are you using _@doc_ to comment out parts of your HTML?  Don't do that.  Just use _@_ and _@c_.  _@_ will comment out everything until the  next _@c_.  For example, the following will only display the _\<h1>_  and _\<h3>_ lines but not the _\<h2>_ line when the html file is opened in a browser:

```
@language xml
<html>
<h1>hi there</h1>
@
<h2>2nd level heading</h2>
@c
<h3> 3rd level heading</h3>
</html>
```
This doesn't work for _@language html_ nodes because the entire node content is just sent as is to the Qt browser component in VR3.

**Comment by tbpassin** [comment]
If you have html masquerading as xml, you can render the markup in VR3 using the  minibuffer command `vr3-render-html-from-clip`.  To make this work, copy everything in the VR3 panel to the clipboard, then issue the command.

This way, the effects of the _@_/_@c_ pairs will be honored, and you won't have to open the external file in a web browser.

**Comment by wangzhaohe** [comment]
> 

Yes, that is great. It provide a method to hide document by @&@c.

And show the document by @doc&@code.

**Comment by wangzhaohe** [comment]
> Oh, maybe I see what you are thinking of. Are you using _@doc_ to comment out parts of your HTML? Don't do that. Just use _@_ and _@c_. _@_ will comment out everything until the next _@c_. For example, the following will only display the _<h1>_ and _<h3>_ lines but not the _<h2>_ line when the html file is opened in a browser:
> 
> ```
> @language xml
> <html>
> <h1>hi there</h1>
> @
> <h2>2nd level heading</h2>
> @c
> <h3> 3rd level heading</h3>
> </html>
> ```
> 
> This doesn't work for _@language html_ nodes because the entire node content is just sent as is to the Qt browser component in VR3.

Yes, that is great. It provide a method to hide document by @&@c.

And show the document by @doc&https://github.com/code.

**Comment by wangzhaohe** [comment]
> Is this supposed to be an asciidoc file or an XML file? And what are those _@doc_ directives intended to do? VR3 does not do anything with them.
> 
> VR3's purpose is to show rendered versions of subtrees, including @file trees. If you don't need to see rendered html, you don't need VR3, it seems to me. With _@language xml_ VR3 does show unrendered but colorized text as you say. But so does the body itself.
> 
> If this is an asciidoc file, then the Asciidoctor processor will render the markup the way it is programmed to do, and that is what you will see in VR3.
> 
> To sum up, I don't understand what you are trying to do, what you want to happen, and what is actually happening that is different from what you want.
> 
> Once I know those things, we can think about whether something can be done.

There are two things should be clarified.

First, this issue is about leo body text syntax color, NO relation with VR3 representation.

Second, using @doc or @ is IMPORTANT. Lines written in @file should not be executed with @doc or @.  I not only need @doc or @ to comment out lines in @file, but also need to show the lines between @doc & @code with VR3 plugin.

**Comment by wangzhaohe** [comment]
> I am not sure exactly what you are trying to write here, but your XML is not well-formed, because XML comments cannot be nested. So far as the XML is concerned, if you want to have what look like comments inside comments, you have to use _>_ and _<_ entities (or at least _<_). Also, inside a comment you cannot have the string "--", according the the RxP XML parser.
> 
> You could put unescaped special characters in a CDATA section, but I'm pretty sure that's not what you want.
> 
> Would you please attach a plain text file that contains the XML content that you want? And then explain exactly what you are trying to do here.

I means that the written result by @file not @clean is different.

1. No nested comments, but leo body can not color the syntax. No relation with VR3 plugin.
```
@language html
@doc
Some code
```

2. with nested comments, the file can not be run correctly by browser. But leo body can color the syntax. No relation with VR3 plugin.
```
@doc
@language html
Some code
```

Perhaps it will be clear that NO relation with VR3 plugin here, only a leo body show.

**Comment by wangzhaohe** [comment]
> If you have html masquerading as xml, you can render the markup in VR3 using the minibuffer command `vr3-render-html-from-clip`. To make this work, copy everything in the VR3 panel to the clipboard, then issue the command.
> 
> This way, the effects of the _@_/_@c_ pairs will be honored, and you won't have to open the external file in a web browser.

`vr3-render-html-from-clip` is an excited command. I usually change @language xml to @language html repeatedly.

**Comment by tbpassin** [comment]
> Yes, that is great. It provide a method to hide document by @&@c.
> 
> And show the document by @doc & @code.

BTW, _@c _ means the same as _@code_.  And _@_ is, I think, the same as _@doc_ (at least, if the file contains a single node).

The problem with your example is that you have the string "-->" in the content of that sentence.  That string ends the html comment that was started by _@doc_ or _@_.  With HTML and XML, you need to use `&gt;` instead:
```
<p>with syntax color, but --&gt; will show in html page</p>
```
It may be true that a browser might render a `>` character in text, because they are so tolerant of syntax error, but they cannot do that if it is part of a comment-end string.

You also talk about syntax coloring or the lack of it.  Browsers are not going to display html with syntax coloring.  They are going to display the html that you wrote.  If you are talking about *Leo*'s syntax coloring, you may as well use _@language html_, since that will colorize all right.  I don't see any advantage to using _@language xml_

I notice that within a section switched off by _@_ or _@doc_, Leo's colorizing is switched off.  That seems to make sense to me.  As soon as the _@_ is changed to _@c_, the colorizing is applied again.

To repeat, neither XML nor HTML allow nested comments.  If you have reserved characters or strings in your HTML content, you should replace them with their equivalent entities: `&lt;`, `&gt;`, `&apos`, `&quot;`, and `&amp;`.  [HTML also pre-defines some other entities (which are properly called  "character entity references")].

**Comment by wangzhaohe** [comment]
@tbpassin Yes, it is a common phenomenon, not only with xml/html but also with python/java/javascript/dart etc.

**Comment by edreamleo** [comment]
I am closing this issue. Leo's `@doc` parts are working as expected. So is Leo's syntax colorer.
